### PR TITLE
Migrated to swift and fixed rendering bugs

### PR DESCRIPTION
**Pull request by danielgindi**
I have migrated everything to Swift (with proper ObjC compat).

Along the way I have solved the main rendering issues - the one where too many layer recreations are performed, and the one where stroke line is bleeding outside of the component.

@Boris-Em I think this will make many people happy. Please take a look.

### CODE REVIEW

File: Classes/BEMAnimationManager.swift [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        }
+        animation.duration = animationDuration
+        animation.isRemovedOnCompletion = false
+        animation.fillMode = .forwards
+        animation.timingFunction = CAMediaTimingFunction(name: .easeInEaseOut)
+        
+        return animation
+    }
+    
+    /** Returns a CABasicAnimation which animates the opacity.
```
Comment (Boris-Em) [inline Classes/BEMAnimationManager.swift]:
```suggestion
    /** Returns a `CABasicAnimation` instance which animates the opacity.
```

File: Classes/BEMAnimationManager.swift [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        }
+        animation.duration = animationDuration
+        animation.isRemovedOnCompletion = false
+        animation.fillMode = .forwards
+        animation.timingFunction = CAMediaTimingFunction(name: .easeInEaseOut)
+        
+        return animation
+    }
+    
+    /** Returns a CABasicAnimation which animates between two paths.
```
Comment (Boris-Em) [inline Classes/BEMAnimationManager.swift]:
```suggestion
    /** Returns a `CABasicAnimation` instance which animates between two paths.
```

File: Classes/BEMCheckBox.swift [hunk 3]
```diff
@@ -0,0 +1,621 @@
...
+//  CheckBox
+//
+//  Created by Bobo on 9/19/15.
+//  Copyright (c) 2015 Boris Emorine. All rights reserved.
+//
+
+import UIKit
+
+/** The BEMCheckBoxDelegate protocol. Used to receive life cycle events. */
+@objc public protocol BEMCheckBoxDelegate: NSObjectProtocol {
```
Comment (Boris-Em) [inline Classes/BEMCheckBox.swift]: Same here, are we sure that we need the `@objc` attribute?

File: Classes/BEMCheckBox.swift [hunk 4]
```diff
@@ -0,0 +1,621 @@
...
+        case circle
+        
+        /** Squared box. */
+        case square
+    }
+    
+    @objc(BEMAnimationType)
+    public enum AnimationType : Int {
+        /** Animates the box and the check as if they were drawn.
+         *  Should be used with a clear colored onFillColor property. */
```
Comment (Boris-Em) [inline Classes/BEMCheckBox.swift]:
```suggestion
         *  Should be used with a clear colored `onFillColor` property. */
```

File: Classes/BEMCheckBox.swift [hunk 5]
```diff
@@ -0,0 +1,621 @@
...
+        case stroke
+        
+        /** When tapped, the checkbox is filled from its center.
+         * Should be used with a colored onFillColor property. */
```
Comment (Boris-Em) [inline Classes/BEMCheckBox.swift]:
```suggestion
         * Should be used with a colored `onFillColor` property. */
```

File: Classes/BEMCheckBox.swift [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+        
+        if on {
+            setupCheckmarkLayer()
+        }
+    }
+
+    /** Draws the box used when the checkbox is set to Off. */
+    private func setupBoxOffLayer() {
+        if offBoxLayer == nil
+        {
```
Comment (Boris-Em) [inline Classes/BEMCheckBox.swift]: For consistency, we should open the curly bracket on the same line as the `if`.

File: Classes/BEMCheckBox.swift [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+        offBoxLayer?.lineWidth = lineWidth
+        
+        self.layer.insertSublayer(offBoxLayer!, at: 0)
+    }
+    
+    /** Draws the box when the checkbox is set to On.
+     */
+    private func setupBoxOnLayer() {
+        if onBoxLayer == nil
+        {
```
Comment (Boris-Em) [inline Classes/BEMCheckBox.swift]: For consistency, we should open the curly bracket on the same line as the if.

File: Classes/BEMCheckBox.swift [hunk 8]
```diff
@@ -0,0 +1,621 @@
...
+        onBoxLayer?.strokeColor = onTintColor?.cgColor
+        
+        layer.addSublayer(onBoxLayer!)
+    }
+
+    /** Draws the check mark when the checkbox is set to On.
+     */
+    private func setupCheckmarkLayer() {
+        if checkMarkLayer == nil
+        {
```
Comment (Boris-Em) [inline Classes/BEMCheckBox.swift]: For consistency, we should open the curly bracket on the same line as the if.

File: Classes/BEMAnimationManager.swift [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+//  Created by Bobo on 9/19/15.
+//  Copyright (c) 2015 Boris Emorine. All rights reserved.
+//
+
+import UIKit
+
+public class BEMAnimationManager: NSObject {
+    /** The duration of the animation created by the BEMAnimationManager object.
+     */
+    @objc public var animationDuration: CFTimeInterval = 0.0
```
[the remark this answers is not in the document]
Comment (danielgindi) [reply Classes/BEMAnimationManager.swift]: No user should access `BEMAnimationManager` directly, but I did not have the time to touch the tests, so they are still objc.
As for BEMCheckBox, it's for backwards compatibility. I'm sure a lot of users still have some objc code.

File: Classes/BEMCheckBox.swift (hunk 6 above)
Comment (danielgindi) [reply Classes/BEMCheckBox.swift]: Tried to keep your styling of curly braces in first line, but when I got to solving the old issue with the layers and rendering - I fell back to my healthy habits of nice and clean curlies ;-)

**Comment by danielgindi** [comment]
Btw I think your CI has an outdated Swift version

**Comment by danielgindi** [comment]
Fixed CI

**Comment by danielgindi** [comment]
CI has a communications error, needs to be restarted

**Comment by danielgindi** [comment]
@Boris-Em are you on this?

**Comment by danielgindi** [comment]
@Boris-Em Fixed the optionals issue

**Comment by danielgindi** [comment]
Any eta on a release date so I can stop using a git branch? ;-)

**Pull request merged**